### DISCUSSION TITLE
Insights in support of Duplex Contracts

### DISCUSSION THREAD

**Discussion by Danielku15** [CoreWCF/CoreWCF#398] [Q&A]
Hey everyone. 

For us it would be really important to have support for Duplex Contracts and currently this is a missing feature in CoreWCF. 😭 Considering the fact that CoreWCF is heavily based off the original WCF code, I would assume some good foundation should be there, but not everything was fully ported to make it actually working. 

I would be interested to give it a timebox to fiddle with CoreWCF to add Duplex Channel support and contribute it. It would make our lives a lot easier for transitioning to .net 5/6. 

To get kick started I was hoping some contributors or other devs might have investigated this topic already a bit and can share some insights on what is missing to be ported to get it running.

Also some insights on the general development practices are appreciated. Especially the usual practices on how things are ported from the original WCF code to CoreWCF. Is there somewhere an archive (beside the ReferenceSource of MS) you use as starting point for taking over classes before you refactor them? 

I cannot promise anything that I could really reach the state of a proper contribution but I am really eager to give it a timeboxed attempt. 

Looking forward to some feedback 😉

**Comment by Danielku15** [comment]
Update: I started to dig a bit into this area and with only a very minor adaptation I was able to get a basic duplex channel setup running: 
https://github.com/CoreWCF/CoreWCF/pull/399

But my concerns remain open: I am not sure what else might be missing at this point. So far I can only confirm that the happy path to make duplex calls works now.

**Comment by birojnayak** [comment]
@Danielku15 below is my porting practice... 
1)I have downloaded code to my local from https://github.com/microsoft/referencesource 
2)To enable a scenario, I write a test in WCF(client/server) debug end-to-end mostly on the server side. ( if you don't have symbol for servicmodel, here is one of the link I had followed https://www.youtube.com/watch?v=D7spm6C2j30 , if you know kindly ignore). Once I debug and understood, I pick up class one at a time for CoreWCF server side.
3)To add test, please refer here for client side tests https://github.com/dotnet/wcf and see which are relevant for server side too, e:g:-  https://github.com/dotnet/wcf/tree/1a291ea424991a184ffafd2ba20928e0e339fb96/src/System.ServiceModel.Primitives/tests/ServiceModel and start porting one at a time

Also, when the work is big, you can split into multiple PR's (logical division) and address them. I will work @mconnew to get formal documented on above and other things we should consider.
